### PR TITLE
Update docs to highlight lambda comparison

### PR DESCRIPTION
**Pull request by RazvanN7**
For : https://github.com/dlang/dmd/pull/7484

### CODE REVIEW

File: spec/traits.dd [hunk 1]
```diff
@@ -1020,6 +1020,59 @@ void main()
...
+        instantiation as an explicit argument type. Any other argument
+        types (basic, user-defined, template) are supported.)
+        $(LI the lambda function body must contain a single expression
+        (no return statement) which contains only numeric values,
+        manifest constants, enum values and function arguments. If the
+        expression contains local variables, function calls or return
+        statements, the function is considered uncomparable.)
+        )
+
+---
```
Comment (wilzbach) [inline spec/traits.dd]: Try the new `SPEC_RUNNALE_EXAMPLE_COMPILE` macro
Verify with `pmake test_dspec`

Example: https://github.com/dlang/dlang.org/blob/47bfef14e4275f42b403d75b31628d563f57b1d4/spec/type.dd

File: spec/traits.dd [hunk 2]
```diff
@@ -1020,6 +1020,59 @@ void main()
...
+void main()
+{
+    static assert(__traits(isSame, (a, b) => a + b, (c, d) => c + d));
+    static assert(__traits(isSame, a => ++a, b => ++b));
+    static assert(!__traits(isSame, (int a, int b) => a + b, (a, b) => a + b));
+    static assert(__traits(isSame, (a, b) => a + b + 10, (c, d) => c + d + 10));
```
Comment (wilzbach) [inline spec/traits.dd]: FYI people often add one whitespace character here in such cases to keep everything neatly aligned.

File: spec/traits.dd [hunk 3]
```diff
@@ -1020,6 +1020,59 @@ void main()
...
+        $(OL
+        $(LI The lambda function arguments must not have a template
+        instantiation as an explicit argument type. Any other argument
+        types (basic, user-defined, template) are supported.)
+        $(LI the lambda function body must contain a single expression
+        (no return statement) which contains only numeric values,
+        manifest constants, enum values and function arguments. If the
+        expression contains local variables, function calls or return
+        statements, the function is considered uncomparable.)
+        )
```
Comment (wilzbach) [inline spec/traits.dd]: How about adding:

> If these constraints aren't fulfilled,  the function is considered uncomparable and `isSame` returns `false`

File: spec/traits.dd [hunk 4]
```diff
@@ -1020,6 +1020,63 @@ void main()
...
+        $(LI The lambda function arguments must not have a template
+        instantiation as an explicit argument type. Any other argument
+        types (basic, user-defined, template) are supported.)
+        $(LI The lambda function body must contain a single expression
+        (no return statement) which contains only numeric values,
+        manifest constants, enum values and function arguments. If the
+        expression contains local variables, function calls or return
+        statements, the function is considered uncomparable.)
+        $(LI If these constraints aren't fulfilled, the function is considered
+        uncomparable and `isSame` returns $(D false).)
```
Comment (wilzbach) [inline spec/traits.dd]: This is not part of the list and should be put in a new paragraph after the list.

File: spec/traits.dd [hunk 5]
```diff
@@ -1020,6 +1020,63 @@ void main()
...
+
+$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+---
+void main()
+{
+    static assert(__traits(isSame, (a, b) => a + b, (c, d) => c + d));
+    static assert(__traits(isSame, a => ++a, b => ++b));
+    static assert(!__traits(isSame, (int a, int b) => a + b, (a, b) => a + b));
+    static assert(__traits(isSame, (a, b) => a + b + 10, (c, d) => c + d + 10));
+
```
Comment (wilzbach) [inline spec/traits.dd]: // lambdas accessing local variables are considered incomparable

File: spec/traits.dd [hunk 6]
```diff
@@ -1020,6 +1020,63 @@ void main()
...
+    int b;
+    static assert(!__traits(isSame, a => a + b, a => a + b));
+
```
Comment (wilzbach) [inline spec/traits.dd]: // lambas calling local functions are considered incomparable.

File: spec/traits.dd (hunk 4 above)
Comment (wilzbach) [inline spec/traits.dd]: incomparable!

File: spec/traits.dd [hunk 7]
```diff
@@ -1020,6 +1020,63 @@ void main()
...
+    class B
+    {
+        int a;
+        this(int a)
+        {
+            this.a = a;
+        }
+    }
+
+    static assert(__traits(isSame, (A a) => ++a.a, (A b) => ++b.a));
```
Comment (wilzbach) [inline spec/traits.dd]: // lambdas with different data types are incomparable - even if have the same memory layout

**Comment by RazvanN7** [comment]
@wilzbach Done

**Comment by ZombineDev** [comment]
@RazvanN7 squash please.

**Comment by RazvanN7** [comment]
@wilzbach @ZombineDev sorry for the late response. I rebased. Thanks!

**Pull request merged**